plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -521,10 +524,28 @@ private IcebergSplit toIcebergSplit(FileScanTask task, TupleDomain<IcebergColumn
                 task.deletes().stream()
                         .map(DeleteFile::fromIceberg)
                         .collect(toImmutableList()),
-                SplitWeight.fromProportion(clamp((double) task.length() / tableScan.targetSplitSize(), minimumAssignedSplitWeight, 1.0)),
+                SplitWeight.fromProportion(clamp(getSplitWeight(task), minimumAssignedSplitWeight, 1.0)),
                 fileStatisticsDomain,
                 fileIoProperties,
                 cachingHostAddressProvider.getHosts(task.file().path().toString(), ImmutableList.of()),
                 task.file().dataSequenceNumber());
     }
+
+    private double getSplitWeight(FileScanTask task)
+    {
+        double dataWeight = (double) task.length() / tableScan.targetSplitSize();
+        double weight = dataWeight;
+        if (task.deletes().stream().anyMatch(deleteFile -> deleteFile.content() == POSITION_DELETES)) {
+            // Presence of each data position is looked up in a combined bitmap of deleted positions
+            weight += dataWeight;
```

Comment:
It does seem heavy, but since the calculation is clamped to `[minimumAssignedSplitWeight, 1.0]`, this really only matters if the split is small in the first place. In practice, this will turn a split that would be "half-weighted" by the data size to be "whole-weighted".

As a worked example, assuming the default value of 128MiB for `read.split.target-size`, this would make a 64MiB scan with positional deletes to be considered to be equal to a 128MiB scan without them- or a 32MiB scan with them equal to a 64MiB without them.

Reply:
Length of delete file wouldn't work well because compression/encoding will reduce it to a small size. The presence of any delete position will result in significant extra work: read delete file, convert to roaring bitmap, produce row positions block, look-up every row position in bitmap, produce a filtered page.